File: Crio/crio_programming_interview_problems-master/RegularExpressionMatching/RegularExpressionMatching.cpp
#include <bits/stdc++.h>
using namespace std;

string regularExpressionMatching(string s, string p)
{
// CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    int m = s.size(), n = p.size();
    string x = "";
    x+=p[0];
    for(int i=1 ;i<n;i++)
    {
        if(x.back() != p[i] or p[i] != '*')
            x+=p[i];
    }
    int pos = 0;
    p = x;
    n = p.size();
    while( pos < n and p[pos] == '*')
        pos+=1;
    
    vector<vector<bool > > dp( m+1 ,vector<bool>(n+1));
    dp[m][n]=true;
    for (int i = m; i >= 0; i--){
        for (int j = n - 1; j >= 0; j--){
            bool first_match = (i < m && (p[j] == s[i] || p[j] == '.'));
            if (j + 1 < n && p[j+1] == '*'){
                dp[i][j] = dp[i][j+2] || first_match && dp[i+1][j];
            } else {
                dp[i][j] = first_match && dp[i+1][j+1];
            }
        }
    }
    if (dp[0][pos])
        return "true";
    else
        return "false";
// CRIO_SOLUTION_END_MODULE_L1_PROBLEMS
}

int main()
{
    string s, p;
    cin>>s>>p;
    string result = regularExpressionMatching(s , p);
    cout<<result<<"\n";

    return 0;
}

File: Crio/crio_programming_interview_problems-master/RegularExpressionMatching/RegularExpressionMatching.py

def regularExpressionMatching(text, pattern):
# CRIO_SOLUTION_START_MODULE_L1_PROBLEMS
    pat = pattern[0]
    for i in pattern[1:]:
        if pat[-1] != i or i != "*" :
            pat+=i
    pattern = pat
    pos = 0
    while pos < len(pattern) and pattern[pos] == "*":
        pos+=1
    n , m = len(text) , len(pattern)
    dp = [[False] * (m + 1) for _ in range(n + 1)]
    dp[-1][-1] = True
    for i in range(n, -1, -1):
        for j in range(m - 1, -1, -1):
            first_match = i < n and pattern[j] in {text[i], '.'}
            if j+1 < m and pattern[j+1] == '*':
                dp[i][j] = dp[i][j+2] or first_match and dp[i+1][j]
            else:
                dp[i][j] = first_match and dp[i+1][j+1]

    return "true" if dp[0][pos] else "false"
# CRIO_SOLUTION_END_MODULE_L1_PROBLEMS

def main():
    s = input().rstrip()
    p = input().rstrip()
    ans  = regularExpressionMatching(s,p)
    print(ans)

if __name__ == "__main__":
    main()
